test/Sema/enum_conformance_synthesis.swift

```diff
@@ -58,10 +58,10 @@ func customHashable() {
 // We still synthesize conforming overloads of '==' and 'hashValue' if
 // explicit definitions don't satisfy the protocol requirements. Probably
 // not what we actually want.
-enum InvalidCustomHashable {
+enum InvalidCustomHashable { // expected-error {{invalid redeclaration of synthesized property 'hashValue'}}
   case A, B
 
-  var hashValue: String { return "" }
+  var hashValue: String { return "" } // expected-note {{'hashValue' previously declared here}}
```

Comment:
This would be confusing to me as a new user. This is the invalid redeclaration; the "previously" declared `hashValue` is invisible because it's synthesized.

I like the new diagnostic, but is there any way to suppress these notes? Every single one points to a location that's not where the previous declaration is.

Reply:
This note used to be there before the regression... now it's restored. Anyway, the reason this is diagnosed as "previous" is because the synthesised `hashValue` declaration appears first in the AST. During redeclaration checking, we look at the "current" declaration (in this case, the synthesised `hashValue`) and find "other" declarations (such as this user-defined `hashValue`) and then diagnose. I suppose we can fix this if the synthesised `hashValue` appeared last in the AST for the enclosed enum.

Reply:
> is there any way to suppress these notes?

Hmm, I suppose we can do that. We can just get rid of the note if the current declaration is implicit and emit the error directly on the other declaration. So, for the above scenario, we'd just have an error on the user-defined `hashValue` i.e. `invalid redeclaration of synthesized property 'hashValue'`. Does that sound reasonable?

Reply:
Suppressing the note seems fine to me, but it would also be super nice to emit a note about where the synthesized property came from, e.g., a note pointing to a property with an attached property wrapper, or a conformance which is synthesized (with better wording than this):

```
struct TestInvalidRedeclaration2 {
   @WrapperWithInitialValue var foo1 = 123 // note: '_foo1' synthesized for property wrapper backing storage

   var _foo1 = 123 // error: invalid redeclaration of synthesized property '_foo1'
 }
```

Reply:
Sure! I think I can add a specialised note for property wrappers.

Reply:
Okay, I have updated the diagnostics. How does it look now @hborla?